src/core/common/trickle_timer.cpp

```diff
@@ -0,0 +1,191 @@
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the trickle timer logic.
+ */
+
+#include <common/code_utils.hpp>
+#include <common/debug.hpp>
+#include <common/trickle_timer.hpp>
+#include <platform/random.h>
+
+namespace Thread {
+
+TrickleTimer::TrickleTimer(
+    TimerScheduler &aScheduler, uint32_t aRedundancyConstant, TrickleTimerMode aMode,
+    Handler aTransmitHandler, Handler aIntervalExpiredHandler, void *aContext)
+    :
+    mTimer(aScheduler, HandleTimerFired, this),
+    k(aRedundancyConstant),
+    mMode(aMode),
+    mPhase(kTricklePhaseDormant),
+    mTransmitHandler(aTransmitHandler),
+    mIntervalExpiredHandler(aIntervalExpiredHandler),
+    mContext(aContext)
+{
+}
+
+bool TrickleTimer::IsRunning(void) const
+{
+    return mPhase != kTricklePhaseDormant;
+}
+
+void TrickleTimer::Start(uint32_t aIntervalMin, uint32_t aIntervalMax)
+{
+    assert(!IsRunning());
+
+    // Set the interval limits
+    Imin = aIntervalMin;
+    Imax = aIntervalMax;
+
+    // Initialize I to [Imin, Imax]
+    I = Imin + otPlatRandomGet() % (Imax - Imin);
+
+    // Start a new interval
+    StartNewInterval();
+}
+
+void TrickleTimer::Stop(void)
+{
+    mPhase = kTricklePhaseDormant;
+    mTimer.Stop();
+}
+
+void TrickleTimer::IndicateConsistent(void)
+{
+    // Increment counter
+    c++;
+}
+
+void TrickleTimer::IndicateInconsistent(void)
+{
+    // Only relavent if we aren't already at 'I' == 'Imin'
+    if (IsRunning() && I != Imin)
+    {
+        // Reset I to Imin
+        I = Imin;
+
+        // Stop the existing timer
+        mTimer.Stop();
+
+        // Start a new interval
+        StartNewInterval();
+    }
+}
+
+void TrickleTimer::StartNewInterval(void)
+{
+    // Reset the counter and timer phase
+    c = 0;
+    mPhase = kTricklePhaseTransmit;
+
+    // Initialize t
+    if (I == 0)
+    {
+        // Immediate interval, just set t to 0
+        t = 0;
+    }
+    else if (mMode == kTrickleTimerModeMPL)
+    {
+        // Initialize t to random value between (0, I]
+        t = otPlatRandomGet() % I;
+    }
+    else
+    {
+        // Initialize t to random value between (I/2, I]
+        t = (I / 2) + otPlatRandomGet() % (I / 2);
+    }
+
+    // Start the timer for 't' milliseconds from now
+    mTimer.Start(t);
+}
+
+void TrickleTimer::HandleTimerFired(void *aContext)
+{
+    TrickleTimer *obj = reinterpret_cast<TrickleTimer *>(aContext);
+    obj->HandleTimerFired();
+}
+
+void TrickleTimer::HandleTimerFired(void)
+{
+    bool ShouldContinue = true;
```

Comment:
nit: use lower camel case, upper camel case is reserved for class, namespace, structure, method, function, enumeration, and type names.